GC: wasmparser: Add user defined array types

imikushin opened the pull request:
This adds basic support for user defined array types, for example:
```wat
  (type $a (array i32))
  (type $b (array (mut i32)))
  (type $c (array (mut (ref null $b))))
```

imikushin commented:
@fitzgen Can you take a look?

imikushin commented:
Weird, CI/Rustfmt check is complaining but locally `cargo fmt` doesn't produce a diff 🤷‍♂️

alexcrichton commented on crates/wasmparser/src/validator/core.rs:
```diff
@@ -513,6 +513,11 @@ impl Module {
                 Type::Func(t)
             }
+            crate::Type::Array(t) => {
+                // TODO check element type. Currently the check panics if the ref is not to a function type.
+                // self.check_revalue_type(t.element_type, features, types, offset)?;
```
A reminder to handle this as it's probably one of the more important parts to handle before landing. (might be good to ensure there's `*.wast` tests for the various invalid cases too)

imikushin replied:
Fixed.

imikushin commented on crates/wasmparser/src/readers/core/types.rs:
```diff
@@ -530,8 +548,10 @@ impl fmt::Display for RefType {
 pub enum HeapType {
-    /// Function of the type at the given index.
+    /// User defined type at the given index.
     TypedFunc(u32),
```
Absolutely. I was thinking of doing a follow-up PR focusing specifically on that. Why not now? It would touch so much code, so it'd be hard to tell that the purpose of the PR is actually what it says in the title 😅

fitzgen replied:
Yeah, this is fine to do in the next PR, since so much stuff gets pulled into the renaming and needs to be adjusted.

imikushin replied:
Added a TODO item for the follow-up.

alexcrichton replied:
FWIW in the future I would personally request that the scattered `// TODO` comments not get committed with PRs. They have a way of not aging well over time, so I'd prefer if things like that are tracked in issues/notes/etc. (this is also one reason I'm more of a fan of big PRs that are complete rather than piecemeal PRs, but that's ok). It's ok to leave these in here for now though.

imikushin commented on crates/wasm-mutate/src/module.rs:
```diff
@@ -66,6 +66,7 @@ impl TryFrom<wasmparser::Type> for TypeInfo {
                     .map(|&t| PrimitiveTypeInfo::from(t))
                     .collect(),
             })),
+            _ => unimplemented!("Array and struct types are not supported yet."),
```
Fixed.

imikushin commented on tests/roundtrip.rs:
```diff
@@ -152,7 +152,12 @@ fn skip_test(test: &Path, contents: &[u8]) -> bool {
     if test.iter().any(|p| p == "gc") {
         // selectively enable some tests
-        let implemented = &["gc-i31.wat", "gc-heaptypes.wat"];
+        let implemented = &[
+            "gc-i31.wat",
+            "gc-heaptypes.wat",
+            "gc-array-types.wat",
+            "gc-array-types-invalid.wast",
+        ];
```
The idea was to enable them selectively out of overall deactivated `gc` directory. I'm not sure if fully enabling all of `gc` is a good idea right now, but I'll try.

imikushin replied:
Done.

imikushin commented on crates/wasmparser/src/validator/types.rs:
```diff
@@ -268,6 +278,7 @@ impl Type {
     pub(crate) fn type_size(&self) -> u32 {
         match self {
             Self::Func(ty) => 1 + (ty.params().len() + ty.results().len()) as u32,
+            Self::Array(_) => 2, // 2 is a guess.
```
Will this also apply to other types? For example, Func: each of the params may vary in payload size.

alexcrichton replied:
It will, yeah.

imikushin replied:
Will address in a follow-up.

imikushin replied:
Filed #1036.

imikushin commented on tests/snapshots/local/gc/gc-array-types.wat.print:
```diff
@@ -0,0 +1,5 @@
+(module
+  (type $a (;0;) (array i32))
+  (type $b (;1;) (array (mut i32)))
+  (type $c (;2;) (array (mut (ref null 1))))
```
That's exactly right. I have seen other examples of `(ref <number>)` and `(func <number>)` all over the place (like `tests/snapshots/local/function-references/call_ref/call_ref.wast/0.print`), so thought that was the desired outcome.

Besides, how would wasmprinter know the name to print: `$a` or `$b` (or any other name)?

alexcrichton replied:
State is tracked for the name of each index, and there's various methods for printing either a number or the name. Something is printing just the number, but it should instead be calling the equivalent "print the named index of this type" method (I forget the name off the top of my head)

imikushin replied:
Filed #1034 to address this.

alexcrichton commented on crates/wasmparser/src/readers/core/types.rs:
```diff
@@ -340,10 +341,27 @@ impl RefType {
+            HeapType::Func => Self::FUNC_KIND,
+            HeapType::Array => Self::ARRAY_KIND,
+            HeapType::Struct => Self::STRUCT_KIND,
+            _ => {
+                return None;
+            }
```
Perhaps? I'm honestly not sure. I find it really difficult to review piecemeal incremental patches like this since I don't have the whole picture in my head. I forget what the previous version of this function now but it felt wrong to pass in `TypedFunc` but get out an indexed array (or something like that). Even now returning `None` with a catch-all `_` feels wrong here since that should probably indicate a programmer error than anything else. Additionally this is where I'm not sure that the design of `HeapType` makes sense to take as an argument here because it already has an indexed variant but it's explicitly not used here.

Overall I'm not sure how best to leave this. If it sticks with `indexed(...)` then I think it should take the `Self::*_KIND` as an argument rather than `HeapType`. Otherwise I'm not sure how best to organize this.

imikushin replied:
Done the refactoring with separate functions for each kind of indexed ref.

alexcrichton commented on crates/wasmparser/src/validator/core.rs:
```diff
@@ -817,7 +813,7 @@ impl Module {
             HeapType::TypedFunc(type_index) => {
                 // Just check that the index is valid
-                self.func_type_at(type_index.into(), types, offset)?;
+                self.type_at(type_index, offset)?;
```
Reading over this again I'm not sure that this is correct? This is verifying that the index is valid but shouldn't the index be further verified? 

I'm not sure if this falls out from type-checking later on or whether it's required happen here. Given that `TypedFunc` should be spelled `Indexed` there's technically not enough context here to verify what the type index should be, so maybe this is ok? I'm not sure if other type-checking contexts are prepared for this validation to no longer happen though.

alexcrichton replied:
In any case function-references and gc are both behind off-by-default flags, so this can be figured out later if you're unable to bottom it out now.

imikushin replied:
Filed #1039.

alexcrichton commented:
I'm a bit strapped for time this week so I'm going to hand this over to @fitzgen from here.

imikushin commented on crates/wasmparser/src/validator/core.rs:
```diff
@@ -802,7 +806,7 @@ impl Module {
     }
 
-    fn check_ref_type(&self, ty: RefType, types: &TypeList, offset: usize) -> Result<()> {
+    fn check_ref_type(&self, ty: RefType, _types: &TypeList, offset: usize) -> Result<()> {
```
Done.

imikushin commented on crates/wasmparser/src/readers/core/types.rs:
```diff
@@ -801,3 +833,14 @@ impl<'a> FromReader<'a> for FuncType {
+        let mutable = reader.read_u8()?;
+        Ok(ArrayType {
+            element_type,
+            mutable: (mutable != 0x00),
```
Done.

fitzgen reviewed (commented):
A few things to clean up and/or file follow up issues for before we merge this. Thanks!

fitzgen commented on crates/wasm-encoder/src/core/types.rs:
```diff
@@ -201,6 +201,18 @@ impl TypeSection {
     }
+
+    /// Define an array type in this type section.
+    pub fn array<T>(&mut self, ty: T, mutable: bool) -> &mut Self
+    where
+        T: Into<ValType>,
```
Nevermind, I forgot that `StorageType` is the next PR that is on top of this.

fitzgen commented:
@imikushin: I went ahead and resolved all the comment threads that are, well, resolved. Anything that is still open should correspond to either something that needs to change in this PR or a follow up issue that needs to be filed.

imikushin commented on crates/wasmparser/src/readers/core/types.rs:
````diff
@@ -166,10 +166,11 @@ impl fmt::Display for ValType {
 //   0000 = none
 //   ```
 // - `(unused)` is unused sequence of bits
+// TODO (unused) is supposed to be zeroed out
````
Removing the TODO. Filed #1037 instead.

imikushin commented on crates/wasmparser/src/readers/core/types.rs:
```diff
@@ -339,11 +340,44 @@ impl RefType {
+    /// limits and therefore is not representable, or when the heap type is not
+    /// a typed array, struct or function.
+    const fn indexed(nullable: bool, heap_type: HeapType, index: u32) -> Option<Self> {
+        let kind = match heap_type {
+            HeapType::Func => Self::FUNC_KIND,
+            HeapType::Array => Self::ARRAY_KIND,
+            HeapType::Struct => Self::STRUCT_KIND,
+            _ => {
+                return None;
+            }
+        };
```
Thanks! This _is_ much cleaner.

imikushin commented on crates/wasm-mutate/src/mutators/add_type.rs:
```diff
@@ -72,6 +72,9 @@ impl Mutator for AddTypeMutator {
                         types.function(params, results);
                     }
+                    wasmparser::Type::Array(_) => {
+                        unimplemented!("Array and struct types are not supported yet.")
```
Done.

imikushin commented:
> @imikushin: I went ahead and resolved all the comment threads that are, well, resolved. Anything that is still open should correspond to either something that needs to change in this PR or a follow up issue that needs to be filed.

@fitzgen Thanks. Addressed the feedback, PTAL.

imikushin commented:
@alxgrk Not yet. I'm working on it though, so in about a month it will.

fitzgen reviewed (approved):
Thanks!
